### ISSUE TITLE
Websocket example is not working

### ISSUE THREAD

**Issue by mcorbin** [sunng87/ring-jetty9-adapter#134]
Hello,

I'm currently migrating one of my project from an old version of ring jetty9 adapter (before the ring websocket standardization) to the latest lib version.

I noticed that the websocket [example provided here](https://github.com/sunng87/ring-jetty9-adapter/blob/master/examples/rj9a/websocket.clj) is not working on ping requests.

When running the example or my own project, I have this error on ping requests:

```
{"@timestamp":"2024-04-18T21:35:45.978+02:00","@version":"1","message":"Error in the websocket handler","logger_name":"mirabelle.transport.websocket","thread_name":"qtp1921836051-35","level":"ERROR","level_value":40000,"stack_trace":"org.eclipse.jetty.websocket.core.exception.WebSocketException: websocket$proxy_ws_adapter$reify__630 PING method error: class clojure.lang.PersistentArrayMap cannot be cast to class java.lang.Class (clojure.lang.PersistentArrayMap is in unnamed module of loader 'app'; java.lang.Class is in module java.base of loader 'bootstrap')
\tat org.eclipse.jetty.websocket.common.JettyWebSocketFrameHandler.onPingFrame(JettyWebSocketFrameHandler.java:312)
\tat org.eclipse.jetty.websocket.common.JettyWebSocketFrameHandler.onFrame(JettyWebSocketFrameHandler.java:214)
\tat org.eclipse.jetty.websocket.core.WebSocketCoreSession$IncomingAdaptor.lambda$onFrame$0(WebSocketCoreSession.java:647)
\tat org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.run(ContextHandler.java:1298)
\tat org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.run(ContextHandler.java:1285)
\tat org.eclipse.jetty.websocket.core.server.internal.AbstractHandshaker$1.handle(AbstractHandshaker.java:179)
\tat org.eclipse.jetty.websocket.core.WebSocketCoreSession$IncomingAdaptor.onFrame(WebSocketCoreSession.java:647)
\tat org.eclipse.jetty.websocket.core.AbstractExtension.nextIncomingFrame(AbstractExtension.java:145)
\tat org.eclipse.jetty.websocket.core.internal.PerMessageDeflateExtension.nextIncomingFrame(PerMessageDeflateExtension.java:239)
\tat org.eclipse.jetty.websocket.core.util.DemandingFlusher.emitFrame(DemandingFlusher.java:143)
\tat org.eclipse.jetty.websocket.core.internal.PerMessageDeflateExtension$IncomingFlusher.handle(PerMessageDeflateExtension.java:382)
\tat org.eclipse.jetty.websocket.core.util.DemandingFlusher.process(DemandingFlusher.java:167)
\tat org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:250)
\tat org.eclipse.jetty.util.IteratingCallback.succeeded(IteratingCallback.java:377)
\tat org.eclipse.jetty.websocket.core.util.DemandingFlusher.onFrame(DemandingFlusher.java:105)
\tat org.eclipse.jetty.websocket.core.internal.PerMessageDeflateExtension.onFrame(PerMessageDeflateExtension.java:96)
\tat org.eclipse.jetty.websocket.core.ExtensionStack.onFrame(ExtensionStack.java:113)
\tat org.eclipse.jetty.websocket.core.WebSocketCoreSession.onFrame(WebSocketCoreSession.java:463)
\tat org.eclipse.jetty.websocket.core.WebSocketConnection.onFrame(WebSocketConnection.java:254)
\tat org.eclipse.jetty.websocket.core.WebSocketConnection.fillAndParse(WebSocketConnection.java:447)
\tat org.eclipse.jetty.websocket.core.WebSocketConnection.onFillable(WebSocketConnection.java:332)
\tat org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
\tat org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
\tat org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
\tat org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:478)
\tat org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:441)
\tat org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
\tat org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produce(AdaptiveExecutionStrategy.java:195)
\tat org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:979)
\tat org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1209)
\tat org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1164)
\tat java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.ClassCastException: class clojure.lang.PersistentArrayMap cannot be cast to class java.lang.Class (clojure.lang.PersistentArrayMap is in unnamed module of loader 'app'; java.lang.Class is in module java.base of loader 'bootstrap')
\tat clojure.core$implements_QMARK_.invokeStatic(core_deftype.clj:555)
\tat clojure.core$extends_QMARK_.invokeStatic(core_deftype.clj:557)
\tat clojure.core$extends_QMARK_.invoke(core_deftype.clj:557)
\tat ring.adapter.jetty9.websocket$proxy_ws_adapter$reify__630.onWebSocketPing(websocket.clj:63)
\tat org.eclipse.jetty.websocket.common.JettyWebSocketFrameHandler.onPingFrame(JettyWebSocketFrameHandler.java:306)
\t... 31 common frames omitted
"}
```

I looked at the code and it's [this part that is failing](https://github.com/sunng87/ring-jetty9-adapter/blob/master/src/ring/adapter/jetty9/websocket.clj#L63). 

The `listener` variable is a map (it's the value of `:ring.websocket/listener` containing `:on-open` etc to my understanding) and so doing `(extends? ring-ws/PingListener listener)` fails (I can reproduce it on a REPL).

By removing the `when` condition and building the lib locally, I was able to successfully implement my websocket server (ping requests are now working, the code [is here](https://github.com/appclacks/mirabelle/pull/26/files) if you want to take a look).

I'm not sure if removing the  `when` condition is the proper fix or not for this issue, what do you think about it ?

Thank you

**Comment by sunng87** [comment]
Thank you for reporting. It seems that I'm using `extends?` wrong here. `extends?` is for checking **type** while we only have value here. 

I'm afraid we cannot simply remove this check because in new ring spec, ping handler is an optional one and we have to check if user really want handle ping here.

**Comment by mcorbin** [comment]
Could we check that the listener map contains `:on-ping` instead ?

**Comment by sunng87** [comment]
I think we can use `instance?` to check and I'm verifying it. Because the `PingListener` may not be a map if user choose to implement it from a  `defrecord` or `deftype`.

**Comment by sunng87** [comment]
finally figured out it's `satisfies?`, creating pull request

**Comment by mcorbin** [comment]
I tested your PR and I confirm it works, thank you.

**Comment by sunng87** [comment]
I will create a release very soon. Note that the new version uses ring 1.12 and you will need to inlcude `ring/ring-core 1.12.1` explicitly in dependencies to be able to use their websocket apis.

**Comment by sunng87** [comment]
released in 0.33.0
